Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/TicketService.hs

```diff
@@ -594,3 +690,290 @@ postTicketBookingsUpdateSeats _ TicketBookingUpdateSeatsReq {..} = do
+  executeNLockOperations lockKeys $ do
+    ticketBooking <- QTB.findByShortId req.ticketBookingShortId >>= fromMaybeM (TicketBookingNotFound req.ticketBookingShortId.getShortId)
+    ticketBookingServices <- QTBS.findByShortIds $ (.shortId) <$> req.ticketBookingServices
+    ticketService <- QTS.findByIds $ (.ticketServiceId) <$> ticketBookingServices
+    ticketBookingServiceCategory <- QTBSC.findByIds $ concatMap (\serviceReq -> (.id) <$> serviceReq.serviceCategory) req.ticketBookingServices
+    ticketBookingPeopleCategory <- QTBPC.findByIds $ concatMap (\serviceReq -> concatMap (\serviceCategory -> (.id) <$> serviceCategory.peopleCategory) serviceReq.serviceCategory) req.ticketBookingServices
+    let ticketBookingServiceMap = Map.fromList $ (\tbs -> (tbs.shortId, tbs)) <$> ticketBookingServices
+    let ticketServiceMap = Map.fromList $ (\ts -> (ts.id, ts)) <$> ticketService
+    let ticketBookingServiceCategoryMap = Map.fromList $ (\tbsc -> (tbsc.id, tbsc)) <$> ticketBookingServiceCategory
+    let ticketBookingPeopleCategoryMap = Map.fromList $ (\tbpc -> (tbpc.id, tbpc)) <$> ticketBookingPeopleCategory
+    validationResp <- validateCancellationRequest ticketBooking req ticketBookingServiceMap ticketServiceMap ticketBookingServiceCategoryMap ticketBookingPeopleCategoryMap
+
+    resp <- mapM (cancelTicketBookingService ticketBooking) validationResp.cancelledServices
+    let totalRefundAmount = sum $ fst <$> resp
+    let totalNoOfCancelledSeats = sum (snd <$> resp) + fromMaybe 0 ticketBooking.cancelledSeats
+    let status = if totalNoOfCancelledSeats == ticketBooking.bookedSeats then DTTB.Cancelled else ticketBooking.status
+
+    QTB.updateStatusAndCancelledSeatsById status (Just totalNoOfCancelledSeats) ticketBooking.id
+
+    when (totalRefundAmount > 0) $ intializeRefundProcess ticketBooking.shortId (Just ticketBooking.ticketPlaceId) totalRefundAmount merchantId ticketBooking.merchantOperatingCityId
+
+  pure Kernel.Types.APISuccess.Success
+  where
+    makeUserLockKeyTuple :: API.Types.UI.TicketService.TicketBookingCancelReq -> [(Kernel.Types.Id.Id Domain.Types.BusinessHour.BusinessHour, Data.Time.Calendar.Day, Kernel.Types.Id.Id Domain.Types.ServiceCategory.ServiceCategory)]
+    makeUserLockKeyTuple req' = concatMap (\serviceReq -> ((\category -> (serviceReq.businessHourId, category.visitDate, category.serviceCategoryId)) <$> serviceReq.serviceCategory)) req'.ticketBookingServices
+
+    executeNLockOperations :: [(Kernel.Types.Id.Id Domain.Types.BusinessHour.BusinessHour, Data.Time.Calendar.Day, Kernel.Types.Id.Id Domain.Types.ServiceCategory.ServiceCategory)] -> Environment.Flow () -> Environment.Flow ()
+    executeNLockOperations [] cancelAndRefundOperation = cancelAndRefundOperation
+    executeNLockOperations ((businessHourId, visitDate, serviceCategoryId) : lockKeys) cancelAndRefundOperation = do
+      isUserLockAcquired <- tryUserTicketCancellationLock businessHourId visitDate serviceCategoryId
```

Comment:
At once, for a show only one user is allowed to cancel. Can we make this on shortId of ticketBooking?

Reply:
we can do this but I think this will be more safe 
as we are also updating the seat management 
and this will block again refund request from same user twice

Reply:
Why suggested method will be unsafe? The way it is implemented, only one user is allowed to cancel a show at a moment. This is blocking every other user

Reply:
@0utkarsh bro the logic is such that merchant will generate merchant unique lock key from bHourId, visitDate, serviceCategoryId and similarly user will check if this lock key is present if it is present then it will throw an error and if then we it will generate user unique identifier and if merchant cancellation starts then it will check if user lock key is present if true throw error,
@0utkarsh so we can't add shortId of ticketBooking but we can make it such that if another user also comes we will wait for first operation to complete and execute this one, works ?